packages/core/src/memo.ts

```diff
@@ -0,0 +1,29 @@
+import type { StateNode } from '.';
+
+const stateNodeMap = new WeakMap<StateNode, any>();
+
+export function getMemo<T>(
+  stateNode: StateNode<any, any>,
+  key: string
+): T | undefined {
+  return stateNodeMap.get(stateNode)?.[key];
+}
+
+export function memo<T>(
+  stateNode: StateNode<any, any>,
+  key: string,
+  fn: () => T
+): T {
+  let memoizedData = stateNodeMap.get(stateNode);
+
+  if (!memoizedData || !memoizedData[key]) {
```

Comment:
if the intention is to always cache the first run of `fn` then the more appropriate check here would be to check if the key exists in the `memoizedData`. Falsy values should, technically, be cacheable too - even if we don't quite intend to cache such